File: frontend/lib/models/stock.dart
class Stock {
  final String id;
  final String productId;
  final String storeId;
  final int quantity;
  final int minQuantity;
  final bool isActive;
  final DateTime lastUpdated;
  final String? description;
  final String? storeName;

  Stock({
    required this.id,
    required this.productId,
    required this.storeId,
    required this.quantity,
    required this.minQuantity,
    required this.isActive,
    required this.lastUpdated,
    this.description,
    this.storeName,
  });

  factory Stock.fromJson(Map<String, dynamic> json) {
    final product = json['productId'];
    final store = json['storeId'];
    return Stock(
      id: json['_id'] is Map
          ? json['_id']['\$oid'] ?? json['_id'].toString()
          : json['_id'].toString(),
      productId: product is Map
          ? product['_id']?.toString() ?? product.toString()
          : product.toString(),
      storeId: store is Map
          ? store['_id']?.toString() ?? store.toString()
          : store.toString(),
      quantity: json['quantity'] ?? 0,
      minQuantity: json['minQuantity'] ?? 0,
      isActive: json['isActive'] ?? true,
      lastUpdated:
          DateTime.tryParse(json['lastUpdated'] ?? json['updatedAt'] ?? '') ??
          DateTime.now(),
      description: product is Map ? product['description']?.toString() : null,
      storeName: store is Map ? store['name']?.toString() : null,
    );
  }

  Map<String, dynamic> toJson() => {
    '_id': id,
    'productId': productId,
    'storeId': storeId,
    'quantity': quantity,
    'minQuantity': minQuantity,
    'isActive': isActive,
    'lastUpdated': lastUpdated.toIso8601String(),
    'description': description,
    'storeName': storeName,
  };

  String get productName => description ?? productId;
}

File: backend/src/models/Stock.js
import mongoose from 'mongoose';

const stockSchema = new mongoose.Schema({
  productId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Product', 
    required: true 
  },
  storeId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Store', 
    required: true 
  },
  quantity: { 
    type: Number, 
    default: 0,
    min: 0
  },
  reservedQuantity: { 
    type: Number, 
    default: 0,
    min: 0
  }, // Pour les commandes en cours
  lastUpdated: { 
    type: Date, 
    default: Date.now 
  },
  lastMovementId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'StockMovement'
  },
  minQuantity: {
    type: Number,
    default: 0,
    min: 0
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Index composé pour éviter les doublons et optimiser les recherches
stockSchema.index({ productId: 1, storeId: 1 }, { unique: true });
stockSchema.index({ storeId: 1, quantity: 1 });
stockSchema.index({ storeId: 1, lastUpdated: -1 });

// Virtual pour calculer la quantité disponible
stockSchema.virtual('availableQuantity').get(function() {
  return Math.max(0, this.quantity - this.reservedQuantity);
});

// Virtual pour vérifier si le stock est bas
stockSchema.virtual('isLowStock').get(function() {
  // Cette logique sera complétée avec les données du produit
  return this.quantity <= 5; // Valeur par défaut
});

// Virtual pour vérifier si le stock est épuisé
stockSchema.virtual('isOutOfStock').get(function() {
  return this.quantity <= 0;
});

// Méthode pour ajouter du stock
stockSchema.methods.addStock = function(quantity, reason = 'Ajout manuel') {
  if (quantity <= 0) {
    throw new Error('La quantité doit être positive');
  }
  
  this.quantity += quantity;
  this.lastUpdated = new Date();
  
  return this;
};

// Méthode pour retirer du stock
stockSchema.methods.removeStock = function(quantity, reason = 'Retrait manuel') {
  if (quantity <= 0) {
    throw new Error('La quantité doit être positive');
  }
  
  if (this.quantity < quantity) {
    throw new Error('Stock insuffisant');
  }
  
  this.quantity -= quantity;
  this.lastUpdated = new Date();
  
  return this;
};

// Méthode pour réserver du stock
stockSchema.methods.reserveStock = function(quantity) {
  if (quantity <= 0) {
    throw new Error('La quantité doit être positive');
  }
  
  if (this.availableQuantity < quantity) {
    throw new Error('Stock disponible insuffisant');
  }
  
  this.reservedQuantity += quantity;
  this.lastUpdated = new Date();
  
  return this;
};

// Méthode pour libérer du stock réservé
stockSchema.methods.releaseReservedStock = function(quantity) {
  if (quantity <= 0) {
    throw new Error('La quantité doit être positive');
  }
  
  if (this.reservedQuantity < quantity) {
    throw new Error('Stock réservé insuffisant');
  }
  
  this.reservedQuantity -= quantity;
  this.lastUpdated = new Date();
  
  return this;
};

// Méthode pour ajuster le stock
stockSchema.methods.adjustStock = function(newQuantity, reason = 'Ajustement manuel') {
  if (newQuantity < 0) {
    throw new Error('La quantité ne peut pas être négative');
  }
  
  this.quantity = newQuantity;
  this.lastUpdated = new Date();
  
  return this;
};

export default mongoose.model('Stock', stockSchema); 